src/server.rs

```diff
@@ -152,8 +160,146 @@ impl Server {
         found
     }
 
-    pub fn update_friends_list(&mut self, friendslist: Vec<Friend>) {
-        self.friends_list = friendslist;
+    /// Updates friends lists of a user
+    /// Propagates to all other friends lists to ensure two-way lookup possible.
+    /// Only call if friends list was obtained directly from Steam API (i.e. friends list is public)
+    pub fn update_friends_list(&mut self, id: SteamID, friendslist: Vec<Friend>) {
+        self.friends_is_public.insert(id, true);
+
+        let oldfriends = self.friends_lists.insert(id, friendslist.clone());
+
+        // Propagate to all other hashmap entries
+        for friend in friendslist.clone() {
+            match self.friends_lists.get_mut(&friend.steamid) {
+                // Friend's friendlist in memory
+                Some(friends_of_friend) => {
+                    match friends_of_friend.iter().position(|f| f.steamid == id) {
+                        Some(friend_index) => {
+                            // player already in friend's friends list, update friend_since in case it changed.
+                            friends_of_friend[friend_index].friend_since = friend.friend_since;
+                        },
+                        None => {
+                            friends_of_friend.push(Friend { steamid: id, friend_since: friend.friend_since });
+                        }
+                    }
```

Comment:
We probably don't even need to worry about the case where it exists and we update the time since, and it could instead be reduced to just adding them if it doesn't exist, or ignoring otherwise.